handler/src/main/java/io/netty/handler/ssl/AbstractSniHandler.java

```diff
@@ -57,156 +54,153 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) t
         if (!suppressRead && !handshakeFailed) {
             final int writerIndex = in.writerIndex();
             try {
-                loop:
-                for (int i = 0; i < MAX_SSL_RECORDS; i++) {
-                    final int readerIndex = in.readerIndex();
-                    final int readableBytes = writerIndex - readerIndex;
-                    if (readableBytes < SslUtils.SSL_RECORD_HEADER_LENGTH) {
-                        // Not enough data to determine the record type and length.
+                final int readerIndex = in.readerIndex();
+                final int readableBytes = writerIndex - readerIndex;
+                if (readableBytes < SslUtils.SSL_RECORD_HEADER_LENGTH) {
+                    // Not enough data to determine the record type and length.
+                    return;
+                }
+
+                final int command = in.getUnsignedByte(readerIndex);
+
+                // tls, but not handshake command
+                switchStatement:
+                switch (command) {
+                case SslUtils.SSL_CONTENT_TYPE_CHANGE_CIPHER_SPEC:
+                case SslUtils.SSL_CONTENT_TYPE_ALERT:
+                    final int len = SslUtils.getEncryptedPacketLength(in, readerIndex);
+
+                    // Not an SSL/TLS packet
+                    if (len == SslUtils.NOT_ENCRYPTED) {
+                        handshakeFailed = true;
+                        NotSslRecordException e = new NotSslRecordException(
+                                "not an SSL/TLS record: " + ByteBufUtil.hexDump(in));
+                        in.skipBytes(in.readableBytes());
+
+                        SslUtils.notifyHandshakeFailure(ctx, e, true);
+                        throw e;
+                    }
+                    if (len == SslUtils.NOT_ENOUGH_DATA ||
+                        writerIndex - readerIndex - SslUtils.SSL_RECORD_HEADER_LENGTH < len) {
+                        // Not enough data
                         return;
                     }
+                    // SNI can't be present in an ALERT or CHANGE_CIPHER_SPEC record, so we'll fall back and assume
+                    // no SNI is present. Let's let the actual TLS implementation sort this out.
```

Comment:
`AbstractSniHandler` is typically responsible for inserting a `SslHandler` into the pipeline after `SNI`. At this point it is likely that there will be no `SslHandler` in the pipeline. IIUC it is likely at this point that we are out of sync with our peer so should we call some overridable method which by default throws some exception and/or closes the channel? Otherwise the content will remain in `in` and it is not clear how we will make progress.